### ISSUE TITLE
Too deep recursion with Template:syn-saurus

### ISSUE THREAD

**Issue by kristian-clausal** [tatuylonen/wiktextract#894]
Parsing page '今生' I get this on the newest commit; this is using a .db from the newest dumpfile from enwiktionary.

```
今生/Chinese/noun: ERROR: too deep recursion during template expansion at ['今生', 'syn-saurus', '#invoke', '#invoke', 'Lua:saurus:saurus()', 'frame:expandTemplate()', 'col3', '#invoke', '#invoke', 'Lua:columns:display()', 'frame:preprocess()', 'm', '#invoke', '#invoke', 'Lua:links/templates:l_term_t()', 'frame:preprocess()', 'm', '#invoke', '#invoke', 'Lua:links/templates:l_term_t()', 'frame:preprocess()',
```

**Comment by kristian-clausal** [comment]
There has been activity in [Module:links](https://en.wiktionary.org/w/index.php?title=Module:links&action=history) at the start of this month.

**Comment by kristian-clausal** [comment]
Yeah, afaict this is just a bug in a common module that slipped through into the current dumpfile, so I've reverted the dumpfile used in the Kaikki.org regeneration to the last one from 20241001.

Leaving this thread open so that when I see it in six months I can go "oh shit I need to change the dumpfile back to latest".

**Comment by xxyzz** [comment]
It should be fixed in the next 20241101 dump file?

**Comment by kristian-clausal** [comment]
Well, hopefully. It's possible this isn't a bug on the wiktionary side, and it's a bug on our side that only expresses in this dumpfile, but it doesn't cost anything to wait and see in this minor case.

**Comment by xxyzz** [comment]
The changes of some used Lua models were made before but not after Oct 20, I think it's likely a bug in our code. Here is what I found: the code stuck in loop when creating links for the last term in page [Thesaurus:今生](https://en.wiktionary.org/wiki/Thesaurus:今生): `[[这]][[一]][[輩子]]`, the bug could be in Lua code handle links in "col3" template parameters.

**Comment by kristian-clausal** [comment]
There's a difference between the dump files (one dump file works, the other one doesn't), so there's *something* that's changed. It can be a bug on our side, but we can wait until next week for a new dumpfile and test if it works, and if it doesn't then we can try to figure things out on our side.

**Comment by xxyzz** [comment]
XML dumps are paused: https://lists.wikimedia.org/hyperkitty/list/xmldatadumps-l@lists.wikimedia.org/thread/BXWJDPO5QI2QMBCY7HO36ELDCRO6HRM4/

the notice said 20241020 dump files "may have underlying data quality issues", 20241001 files are good.

**Comment by kristian-clausal** [comment]
Well, that explains it then! We'll just have to wait for a good dump.

**Comment by xxyzz** [comment]
They seem to fixed the errors and new dump files can be downloaded now but the infinite loop still exists in 20241101 dump file.

**Comment by kristian-clausal** [comment]
Let's wait for the next dump.

**Comment by xxyzz** [comment]
The code enter the loop in the [iterate_content()](https://en.wiktionary.org/wiki/Module:zh-translit#L-89) function of Module:zh-translit, I suspect it's related with how `frame:preprocess` expands wikitext, `preprocess` is called from `Template:expand`<- `Template:get_arguments` in [Module:template parser](https://en.wiktionary.org/wiki/Module:template_parser). Or it could be something else changes the code path.

The last 2 dump files seem to have the same Lua code as the current code on Wiktionary, if they don't have this bug I'm afraid we have to figure it out...

**Comment by kristian-clausal** [comment]
I was hoping that wasn't the case... Do you want to try to fix it?

**Comment by xxyzz** [comment]
I still don't know how this bug happened exactly but I want to try to find it and fix it if possible. At the moment I could only guess it's related with the "m" templates in page [這](https://en.wiktionary.org/wiki/這)'s "zh-pron" template.

**Comment by xxyzz** [comment]
The loop is also on Wiktionary but MediaWiki's parser could detect the loop and exit the expansion, I made a [fix at here](https://github.com/xxyzz/wikitextprocessor/commit/ed2ae66ab69d33f6c507e04f0745b56b7986cbf2). All tests passed except [test_plusobj_template](https://github.com/tatuylonen/wiktextract/blob/d1aaa63207a67ab8934d41240ff292355d2fc32c/tests/test_en_page.py#L695), because the template name is added twice to `Wtp.expand_stack` when `Wtp.expand()` is called from `template_fn`, very annoying and currently I don't know how to remove this duplicated template name in `expand_stack`.

**Comment by xxyzz** [comment]
Another thing I haven't figure out is why the loop is not ended at this "ERROR: too deep recursion during template expansion" error.

**Comment by xxyzz** [comment]
I pass the test by appending a new "TEMPLATE_FN" to `expand_stack` for template expand functions.

**Comment by kristian-clausal** [comment]
Thank you for taking a look at this.

**Comment by kristian-clausal** [comment]
Almost forgot, switch back to using the `latest` enwiktionary dumpfile for kaikki.org regeneration.

**Comment by xxyzz** [comment]
Wikimedia developer said the bug is fixed and the latest November files are accurate: https://lists.wikimedia.org/hyperkitty/list/xmldatadumps-l@lists.wikimedia.org/message/W6IXKU43RQBQV3GT4CCO3TO4CKIF5DFA/
